### PR TITLE
[fix](Nereids) fix fold constant by be return type mismatched

### PR DESCRIPTION
**Pull request by LiBinfeng-01**
when using fold constant by be, the return type of substring('123456', 1, 3) would changed to be text, which we want it to be 3

### CODE REVIEW

**Comment by LiBinfeng-01** [comment]
run buildall

**Review by englefly — DISMISSED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnBE.java [hunk 1]
```diff
@@ -176,7 +176,11 @@ private static Expression replace(
             if (newChild != child) {
                 hasNewChildren = true;
             }
-            newChildren.add(newChild);
+            if (newChild.getDataType() != child.getDataType()) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnBE.java:179]: should use equals

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnBE.java (hunk 1 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnBE.java]: done

**Comment by LiBinfeng-01** [comment]
run p0

**Comment by LiBinfeng-01** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Review by englefly — APPROVED** [review]

**Pull request merged**